File: elpizo/models/realm.py
import logging
import math

from elpizo.models import geometry
from elpizo.models import record
from elpizo.protos import realm_pb2


logger = logging.getLogger(__name__)


class Realm(record.ProtobufRecord):
  PROTOBUF_TYPE = realm_pb2.Realm
  FIELDS = [
      record.Field("name", record.Scalar),
      record.Field("size", geometry.Vector2),
  ]

  def __init__(self, *args, **kwargs):
    self.regions = None
    super().__init__(*args, **kwargs)

  @property
  def bounds(self):
    return geometry.Rectangle(0, 0, self.size.x, self.size.y)

  def load_intersecting_regions(self, bounds):
    left = max([Region.floor(bounds.left), 0])
    top = max([Region.floor(bounds.top), 0])
    right = min([Region.ceil(bounds.right), self.size.x])
    bottom = min([Region.ceil(bounds.bottom), self.size.y])

    for y in range(top, bottom, Region.SIZE):
      for x in range(left, right, Region.SIZE):
          yield self.regions.load(geometry.Vector2(x, y))

  def is_terrain_passable_by(self, entity):
    if not self.bounds.contains(entity.target_bounds):
      return False

    try:
      regions = list(self.load_intersecting_regions(entity.target_bounds))
    except KeyError:
      return False

    return all(region.is_terrain_passable_by(entity) for region in regions)

  def is_passable_by(self, entity):
    if not self.bounds.contains(entity.target_bounds):
      return False

    try:
      regions = set(self.load_intersecting_regions(entity.target_bounds)) | \
                set(self.load_intersecting_regions(entity.bounds))
    except KeyError:
      return False

    return all(region.is_passable_by(entity) for region in regions)


class Layer(record.ProtobufRecord):
  PROTOBUF_TYPE = realm_pb2.Layer
  FIELDS = [
      record.Field("terrain", record.Scalar),
      record.RepeatedField("tiles", record.Scalar),
  ]


class Region(record.ProtobufRecord):
  PROTOBUF_TYPE = realm_pb2.Region
  FIELDS = [
      record.RepeatedField("layers", Layer),
      record.RepeatedField("passabilities", record.Scalar),
      record.RepeatedField("entity_ids_idx", record.Scalar),
  ]

  SIZE = 16

  @classmethod
  def floor(cls, x):
    return (x // cls.SIZE) * cls.SIZE

  @classmethod
  def ceil(cls, x):
    return math.ceil(x / cls.SIZE) * cls.SIZE

  @property
  def id(self):
    return "{x},{y}".format(x=self.location.x, y=self.location.y)

  @property
  def bounds(self):
    return geometry.Rectangle(self.location.x, self.location.y,
                              self.SIZE, self.SIZE)

  @property
  def client_entities(self):
    return (entity for entity in self.entities if entity.client_visible)

  @id.setter
  def id(self, v):
    if v is None:
      return

    x, y = v.split(",")
    self.location = geometry.Vector2(int(x), int(y))

  def to_public_protobuf(self, realm):
    proto = self.to_protobuf()
    proto.ClearField("entity_ids_idx")
    return proto

  def is_terrain_passable_by(self, entity):
    bounds = entity.target_bounds.offset(self.location.negate())

    for y in range(bounds.top, bounds.bottom):
      for x in range(bounds.left, bounds.right):
        if 0 <= x < Region.SIZE and 0 <= y < Region.SIZE and \
          not ((self.passabilities[y * Region.SIZE + x] >>
              entity.direction) & 0x1):
          return False

    return True

  def is_passable_by(self, entity):
    if not self.is_terrain_passable_by(entity):
      return False

    return not any((target.bounds.intersects(entity.target_bounds) or
                    target.bounds.intersects(entity.bounds)) and
                   not target.is_passable_by(entity)
                   for target in self.entities)

  @property
  def channel(self):
    return ("region", self.realm.id, self.location)


class RealmStore(record.ProtobufStore):
  RECORD_TYPE = Realm
  PROTOBUF_TYPE = realm_pb2.Realm

  def __init__(self, parent, kvs):
    self.parent = parent
    super().__init__(kvs)

  def add(self, realm):
    super().add(realm)
    realm.update(regions=self.parent.make_region_store(realm))

  def expire(self, realm):
    super().expire(realm)
    realm.regions.expire_all()

  def save(self, realm):
    super().save(realm)
    if realm.regions is not None:
      logger.info("Saving all child regions for realm: %s", realm.id)
      realm.regions.save_all()


class RegionStore(record.ProtobufStore):
  RECORD_TYPE = Region
  PROTOBUF_TYPE = realm_pb2.Region

  def __init__(self, realm, entities, kvs):
    self.realm = realm
    self.entities = entities
    super().__init__(kvs)

  def load(self, vec):
    return super().load("{x},{y}".format(x=vec.x, y=vec.y))

  def load_closest(self, location):
    return self.load(location.map(Region.floor))

  def find(self, id):
    region = super().find(id)
    region.update(realm=self.realm,
                  entities={self.entities.load(entity_id)
                            for entity_id in region.entity_ids_idx})
    return region

  def save(self, region):
    region.update(entity_ids_idx=[entity.id for entity in region.entities])
    super().save(region)

  def keys(self):
    # We don't want to use the store's integer coercion.
    for key in self.kvs.keys():
      x, y = key.split(",")
      yield geometry.Vector2(int(x), int(y))

File: client/models/realm.js
module grid from "client/util/grid";
module objects from "client/util/objects";
module geometry from "client/models/geometry";

export class Realm {
  constructor(id, message) {
    this.id = id;
    this.name = message.name;
    this.size = geometry.Vector2.fromProtobuf(message.size);

    this.regions = {};
    this.entities = {};
  }

  addRegion(region) {
    this.regions[region.getKey()] = region;
    region.realm = this;
  }

  removeRegion(region) {
    delete this.regions[region.getKey()];
    delete region.realm;
  }

  getRegionAt(location) {
    var key = [location.x, location.y].join(",");
    return objects.hasOwnProp.call(this.regions, key) ? this.regions[key] : null;
  }

  getAllRegions() {
    return Object.keys(this.regions).map((k) => this.regions[k]);
  }

  getClosestRegionTo(location) {
    return this.getRegionAt(location.map(Region.floor));
  }

  getIntersectingRegions(bounds) {
    var regions = [];

    for (var y = Region.floor(bounds.top);
         y < Region.ceil(bounds.getBottom());
         y += Region.SIZE) {
      for (var x = Region.floor(bounds.left);
           x < Region.ceil(bounds.getRight());
           x += Region.SIZE) {
        regions.push(this.getRegionAt(new geometry.Vector2(x, y)));
      }
    }

    return regions;
  }

  getBounds() {
    return new geometry.Rectangle(0, 0, this.size.x, this.size.y);
  }

  getExtendedBounds() {
    return new geometry.Rectangle(0, 0,
                                  Region.ceil(this.size.x),
                                  Region.ceil(this.size.y));
  }

  isTerrainPassableBy(entity) {
    if (!this.getBounds().contains(entity.getTargetBounds())) {
      return false;
    }

    return this.getIntersectingRegions(entity.getTargetBounds()).some(
        (region) =>
            region !== null && region.isTerrainPassableBy(entity));
  }

  isPassableBy(entity) {
    if (!this.isTerrainPassableBy(entity)) {
      return false;
    }

    if (this.getAllEntities().some((target) =>
        (target.getBounds().intersects(entity.getTargetBounds()) ||
         target.getBounds().intersects(entity.getBounds())) &&
        !target.isPassableBy(entity))) {
      return false;
    }

    return true;
  }

  addEntity(id, entity) {
    entity.realm = this;
    this.entities[id] = entity;
  }

  removeEntity(id) {
    var entity = this.entities[id];
    delete this.entities[id];
    delete entity.realm;
  }

  getEntity(id) {
    return objects.hasOwnProp.call(this.entities, id) ? this.entities[id] : null;
  }

  getAllEntities() {
    return Object.keys(this.entities).map((k) => this.entities[k]);
  }

  update(dt) {
    this.getAllEntities().forEach((entity) => {
      entity.update(dt);
    });
  }
}

export class Region {
  constructor(location, message) {
    this.location = location;
    this.realmId = message.realmId;
    this.layers = message.layers.map((layer) =>
        new Layer(layer));
    this.passabilities = new grid.Grid(Region.SIZE, Region.SIZE,
                                       message.passabilities);
  }

  getKey() {
    return [this.location.x, this.location.y].join(",");
  }

  isTerrainPassableBy(entity) {
    var bounds = entity.getTargetBounds().offset(this.location.negate());

    for (var y = bounds.top; y < bounds.getBottom(); ++y) {
      for (var x = bounds.left; x < bounds.getRight(); ++x) {
        var passability = this.passabilities.getCell(x, y);
        if (passability !== null &&
            !((passability >> entity.direction) & 0x1)) {
          return false;
        }
      }
    }

    return true;
  }

  isPassableBy(entity) {
    return this.isTerrainPassableBy(entity);
  }

  getBounds() {
    return (new geometry.Rectangle(0, 0, Region.SIZE, Region.SIZE))
      .offset(this.location);
  }
}
Region.SIZE = 16;

Region.floor = (x) => Math.floor(x / Region.SIZE) * Region.SIZE;
Region.ceil = (x) => Math.ceil(x / Region.SIZE) * Region.SIZE;

class Layer {
  constructor(message) {
    this.terrain = message.terrain;
    this.tiles = new grid.Grid(Region.SIZE, Region.SIZE, message.tiles);
  }
}
